src/schema-validator.ts

```diff
@@ -0,0 +1,139 @@
+import { openmrsFetch } from '@openmrs/esm-framework';
+import { OHRIFormSchema, ConfigObject } from './api/types';
+
+export const handleFormValidation = async (
+  schema: OHRIFormSchema,
+  configObject: ConfigObject,
+): Promise<Array<Array<Record<string, any>>>> => {
+  const errorsArray = [];
+  const warningsArray = [];
+
+  if (schema) {
+    const asyncTasks = [];
+
+    schema.pages?.forEach(page =>
+      page.sections?.forEach(section =>
+        section.questions?.forEach(question => {
+          asyncTasks.push(
+            handleQuestionValidation(question, errorsArray, warningsArray, configObject),
+            handleAnswerValidation(question, errorsArray),
+          );
+          question.type === 'obsGroup' &&
+            question.questions?.forEach(obsGrpQuestion =>
+              asyncTasks.push(
+                handleQuestionValidation(obsGrpQuestion, errorsArray, configObject, warningsArray),
+                handleAnswerValidation(question, errorsArray),
+              ),
+            );
+        }),
+      ),
+    );
+    await Promise.all(asyncTasks);
+
+    return [errorsArray, warningsArray];
+  }
+};
+
+const handleQuestionValidation = async (conceptObject, errorsArray, warningsArray, configObject) => {
+  const conceptRepresentation =
+    'custom:(uuid,display,datatype,answers,conceptMappings:(conceptReferenceTerm:(conceptSource:(name),code)))';
+
+  const searchRef = conceptObject.questionOptions.concept
+    ? conceptObject.questionOptions.concept
+    : conceptObject.questionOptions.conceptMappings?.length
+    ? conceptObject.questionOptions.conceptMappings
+        ?.map(mapping => {
+          return `${mapping.type}:${mapping.value}`;
+        })
+        .join(',')
+    : '';
+
+  if (searchRef) {
+    try {
+      const { data } = await openmrsFetch(`/ws/rest/v1/concept?references=${searchRef}&v=${conceptRepresentation}`);
```

Comment:
I'm concerned about the fact that we have fused the validation logic with API-specific logic. I advise reusing the existing `useConcepts()` hook. This means the validator becomes a React hook:

```js

// the validator runs if doValidate is set to true
// isValidating shows the validation status
const { errors, warnings, isValidating } = useSchemaValidationResults(doValidate, schema);

```

Alternatively, you can define a `FormSchemaValidator` interface but ensure you fetch the concepts in one API hit with proper separation of concerns.

Reply:
@samuelmale I had a previous implementation of how I understood your statement [here](https://github.com/arodidev/openmrs-esm-form-builder/blob/concept-mappings/src/components/interactive-builder/interactive-builder.component.tsx) , kindly advise if this logic suffices.